GPII-1107 & GPII-1191: Solution registry format updated

kaspermarkus opened the pull request:
Fixed the solutions registry format. Still needs a few tests, but most is ready and could use a sanity check/review

klown commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,194 @@
+    "install": [ ... ],
+    "uninstall": [ ... ],
+    "makeConfigurable": [ ... ],
+    "isRunning": { ... },
```
@kaspermarkus

Nit:  "isRunning" is defined as an object here, but on line 154 (R154), it's an array of objects.

klown commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,194 @@
+        "fileName": "firefox.exe"
+    },
+    {
+       "type": "gpii.processReporter.findProcessesByCommand",
```
@kaspermarkus 
I'm not sure if it matters much, but in my process reporter work, I've set the type to "gpii.processReporter.find", meaning, find the process whose command is given by the "command" property.

klown commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,194 @@
+    },
+    {
+       "type": "gpii.processReporter.findProcessesByCommand",
+       "command": "ps aux | grep 'myprocess | wc -l",
```
@kaspermarkus 
Nit:  missing close quote after myprocess.  I think you mean:
ps aux | grep 'myprocess' | wc -l

kaspermarkus commented:
This should be ready for review now!

klown commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,196 @@
+    },
+    {
+       "type": "gpii.processReporter.find",
+       "command": "ps aux | grep 'myprocess | wc -l'",
```
@kaspermarkus,
A number of comments:
First, I'm not clear the intent of the "command" property here.  Is it temporary and just for testing?  Or is that how you envision the actual gpii.processReporter working?

Compare with what I'm doing in my patch with a "type"/"name" pair (I'm fine with using "command" instead of "name").  The "type" specifies the processReporter function, and the "name" is the name of the process to find.  The stuff at the following link means, "look for a process named 'orca'":
https://github.com/klown/universal/blob/GPII-442/testData/solutions/linux.json#L438

Secondly, I think the single quote characters should be around "myprocess" only, that is:
"ps aux | grep 'myprocess' | wc -l"

Thirdly, that will return 1 even if myprocess is not running since it will match the grep itself.  Here is what I get substituting "orca" for "myprocess", i.e. "ps aux | grep 'orca' | wc -l":
* With orca running, outputs 2. 
* With orca off, outputs 1.

Note that running the command without piping to "wc -l", when orca is not running gives a result of:
"clown     4570  0.0  0.2 113000  2376 pts/0    S+   12:41   0:00 grep --color=auto orca"

amb26 replied:
This is a sign that we might want to introduce some kind of annoying compound syntax that will allow us to AND together some directives as well as ORing them

kaspermarkus replied:
@klown This documentation was actually just meant as examples of how things might look once we get the proper support for process detection - so it was meant to follow your format.. Not sure where I got my examples from :) ... Anyway, I've updated it to be similar to your example from your code... I imagine that we might want to have some more detailed documentation about the available process reporters which we would refer to from the solutions registry documentation.. Anyway - hopefully it looks correct now. (PS: I've removed the second block, cause I couldn't really think up a good sensible example for it right now)

@amb26 I think @klown was just referring to my horrific syntax and linux command line skills.. I think we in all cases so far can detect whether a solution is running by a single directive (or one of multiple single directives)... But I might have missed something

klown commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,196 @@
+        "fileName": "firefox.exe"
+    },
+    {
+       "type": "gpii.processReporter.find",
```
@kaspermarkus,

For both orca, and the gnome-shell magnifier, starting/stopping them involves using gsetttings to flip a setting.  However, only in orca's case does a process (the orca process) actually start/stop.  In contrast, no actual magnifier process is started when setting screen-magnifier-enabled to 'true'.  While a gnome-shell process must be running in order for there to be a magnifier, there is no way gpii is going to kill the gnome-shell process to switch off the magnifier.

Thus, the only way to know if the gnome-shell magnifier is on is to check if gnome-shell is running AND the screen-magnifier-enabled gsetting is true.

To handle this case, I modified the processReporter to have a "findAndCheckSetting" function, which requires the setting information to check as well as the process:
https://github.com/klown/universal/blob/GPII-442/testData/solutions/linux.json#L19

I'm not sure this is exactly the way to go, but it's a start.

amb26 commented on documentation/SolutionsRegistryFormat.md:
````diff
@@ -0,0 +1,196 @@
+```
+
+### contexts
+The `contexts` block describes what the required context is for the solution to run. Currently only one type of context is supported, namely `os`. The context block is **mandatory**.
````
Looks like it is capital OS

amb26 commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,196 @@
+### configure, restore, start and stop
+These four lifecycle blocks have different meanings to the system but has the same format. Their meanings are the following:
+
+* `configure`: Configure the solution with the users setting (eg. on login)
```
e.g.

amb26 commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,196 @@
+
+### update
+
+The `update` block works very similarly to the configure, restore, start and stop blocks. It describes what should happen when the configuration needs to be updated (eg. due to context changes, PCP adjustments, etc).
```
e.g.

amb26 commented on gpii/node_modules/contextManager/src/ContextManager.js:
```diff
@@ -169,6 +169,7 @@
 
         lifecycleManager.update(finalPayload, function (response) {
             if (response.success) {
+                delete finalPayload.lifecycleInstructions;
```
Remind why we do this delete? Will this affect @simonbates work on the untrusted local flowmanager?

kaspermarkus replied:
I dont believe this will affect simon, but I guess this change was from before our talk at last arch. meeting... 

The only reason I delete these is to avoid the 'lifecycleInstructions' block being attached to the sessionState in the next line. When doing this patch I got confused that there was so much crap attached to sessionState - and lifecycleInstructions is never used later on, so there was no reason for having it there.. It is only used in the "finalPayload" as instructions on how to configure the system.. once configured, the applied settings are stored in 'activeConfiguration'

amb26 replied:
I'd add a comment here summarising some of this information - it's hard enough for the reader of the code to capture where they are in such a complex lifecycle. Should explain the last sentence of your comment in particular

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -242,23 +236,49 @@ var gpii = fluid.registerNamespace("gpii");
+            return fluid.promise().resolve();
+        }
+        var sequence = fluid.transform(steps, function (action) {
+            if (typeof(action) === "string") { // settings.*  where * can be configure, restore, update, start or stop
```
Break out this inner loop into its own function - it's got fairly gigantic

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -277,28 +297,60 @@ var gpii = fluid.registerNamespace("gpii");
+            promises.push(that.executeActions(solutionId, solutionRecord, sessionState, key));
+        });
+        var promiseSequence = fluid.promise.sequence(promises);
+        //snapshots is an array of snapshotted settingshandler results (one for each lifecycle action block)
```
space here!

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -277,28 +297,60 @@ var gpii = fluid.registerNamespace("gpii");
+     *
+     * @param solutionId {String} the ID of the solution
+     * @param solutionRecord {Object} is not modified and consists of the solution and setings to be applied
+     * @param sessionState {Object} the object holding the state of the system
```
This list of parameters seems pretty messed up and doesn't agree with the actual argument - also, you have sessionState twice

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -277,28 +297,60 @@ var gpii = fluid.registerNamespace("gpii");
-     * @param solutionRecord {Object} is not modified and consists of the original transformer return (doesn't really appear necessary)
-     * @param isUpdate {boolean} indicates whether we're doing an update - in which case the originalSettings should not be written based on snapshot
+     * it is responsible for saving the settings that are being set (when the saveSnapshot is true) and storing
+     * the list ofapplied solutions to the sessionState
```
of applied

amb26 commented:
Thanks for all the work on this, @kaspermarkus - it's a huge amount of great improvement in a really short time! I've given some light review comments so far, but the changes to LifecycleManager.js look pretty extensive and we should probably make time to talk about them intensively to make sure we understand thm.

amb26 commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,191 @@
+    {
+        "type": "gpii.reporter.fileExists",
+        "fileName": "${{registry}.HKEY_CURRENT_USER\\Software\\Texthelp\\Read&Write10\\InstallPath}\\ReadAndWrite.exe"
+    }, // SEEMS LIKE IMPLICIT OR BETWEEN THESE BLOCKS
```
We should probably remove this "SEEMS LIKE" comment as somewhat confidence-sapping in the reader : P

amb26 commented on documentation/SolutionsRegistryFormat.md:
```diff
@@ -0,0 +1,191 @@
+Used for describing the steps required for installing the application
+
+####uninstall:
+Used for describing the steps
```
Incomplete here

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -190,75 +195,102 @@ var gpii = fluid.registerNamespace("gpii");
     };
 
+    /**
+     * @param action {String} should be either a reference to a settings block, written as settings.<blockName>
```
This docs comment is now hugely out of date. Also should add @return {Promise} describing what is returned

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -277,28 +309,59 @@ var gpii = fluid.registerNamespace("gpii");
+
+    /**
+     * Invoked on "stop" phase. Will run the relevant entries via the executeActions function.
+     * The expected outcome is that solution should be restored to it's original state
```
typo its

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -277,28 +309,59 @@ var gpii = fluid.registerNamespace("gpii");
+     * @param sessionState {Object} the object holding the state of the system
+     * @param lifecycleBlockKeys {Array} Array of ordered strings denoting which lifecycle blocks to
+          run (usually "restore" and/or "stop"
+     * @preturn A promise (sequence of promises) as returned by executeActions
```
typo preturn

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -277,28 +309,59 @@ var gpii = fluid.registerNamespace("gpii");
+          run (usually "restore" and/or "stop"
+     * @preturn A promise (sequence of promises) as returned by executeActions
+     */
+    gpii.lifecycleManager.removeSolution = function (that, solutionId, solutionRecord, sessionState, lifecycleBlockKeys) {
```
Shouldn't this be called "stopSolution"?

kaspermarkus replied:
I opted away from "stopSolution" as this can also be run when only restoring settings (and not running the "stop" block) -- but on the other hand, remove solution is a bit misleading as well... Anyway - I renamed to "stopSolution"

klown replied:
@kaspermarkus,
Although you changed the definition to "stopSolution", the invoker is still called "removeSolution":
https://github.com/kaspermarkus/universal/blob/GPII-1107/gpii/node_modules/lifecycleManager/src/LifecycleManager.js#L72

I discovered this when running the acceptance tests, and it failed with, essentially, "can't find defnintion of 'gpii.lifecyleManager.removeSolution'".

Either that, or I'm lost.

FWIW, I too think stopSolution() is better since removeSolution() connotes "uninstall", and that's not what is meant here.  Is "resetToDefautl()" closer?

kaspermarkus replied:
thanks! Merged it i!

wrt the naming, I think resetToDefault is more precise in terms of what it does, but I like that the name is somewhat close to "applySolution" (telling that those two functions are related) ... but there aren't really any good antonyms of "apply" - other than stop

amb26 commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -190,75 +195,102 @@ var gpii = fluid.registerNamespace("gpii");
+     *     function.
+    */
+    gpii.lifecycleManager.executeSettingsAction = function (that, solutionId, solutionRecord, sessionState, settingsReturn, action) {
+        var settingsHandlerBlockName = action.substring(9);
```
Explain this magic constant by expressing it as "some thing".length

amb26 replied:
Also better to do this in the caller where it seems we already express that the magic string is "settings."

kaspermarkus commented:
@amb26 Sorry for not responding to this any sooner - for some reason, I either didn't get or missed the notification emails for your comments! I've addressed the comments now and it should be ready for another round of review

kaspermarkus commented:
Yay!! Cheers, Mr. Bosmon
